src/lib/retry.go

```diff
@@ -0,0 +1,133 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lib
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+var (
+	// ErrRetryTimeout timeout error for retrying
+	ErrRetryTimeout = errors.New("retry timeout")
+)
+
+// RetryOptions options for the retry functions
+type RetryOptions struct {
+	InitialInterval time.Duration                        // the initial interval for retring after failure, default 100 milliseconds
+	MaxInterval     time.Duration                        // the max interval for retring after failure, default 1 second
+	Timeout         time.Duration                        // the total time before returning if something is wrong, default 1 minute
+	Callback        func(err error, sleep time.Duration) // the callback function for Retry when the f called failed
```

Comment:
This is trivial, but IMO 99% of cases of the callback are for printing a log message.
What you need here is a name of the retry session  so you can print a meaning message.